File: apps/api/app/narrative_core/product_contract/structure_map.py
"""Narrative Structure Map projection DTO (Phase 1D-P).

Projection only — nodes/edges from Asset/Relation versions. No Pattern tables.
Not an ORM model and must not be treated as one.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from app.narrative_core.product_contract.enums import StructureMapViewMode

STRUCTURE_MAP_PROJECTION_SCHEMA = "narrative_structure_map_projection"
STRUCTURE_MAP_DEFAULT_MAX_NODES = 100
STRUCTURE_MAP_DEFAULT_MAX_EDGES = 250

# Explicit: this DTO is NOT backed by a database table.
PATTERN_DTO_HAS_ORM_TABLE = False


@dataclass(frozen=True, slots=True)
class StructureMapNodeDto:
    node_id: str
    node_type: str
    title: str
    asset_id: int | None
    asset_version_id: int | None
    is_canonical: bool
    view_modes: tuple[StructureMapViewMode, ...]
    chapter_range: tuple[int | None, int | None] = (None, None)
    parent_id: str | None = None
    evidence_count: int = 0
    collapsed: bool = False
    searchable_text: str = ""


@dataclass(frozen=True, slots=True)
class StructureMapEdgeDto:
    edge_id: str
    source_node_id: str
    target_node_id: str
    relation_id: int | None
    relation_version_id: int | None
    is_canonical: bool
    edge_type: str
    evidence_count: int = 0
    label: str = ""


@dataclass(frozen=True, slots=True)
class StructureMapFiltersDto:
    view_mode: StructureMapViewMode
    search_query: str = ""
    include_candidates: bool = False
    max_nodes: int = STRUCTURE_MAP_DEFAULT_MAX_NODES
    max_edges: int = STRUCTURE_MAP_DEFAULT_MAX_EDGES
    theme: str = "system"  # system | light | dark


@dataclass(frozen=True, slots=True)
class NarrativeStructureMapProjectionDto:
    book_id: int
    book_snapshot_id: int
    source_run_id: int | None
    projection_version: str
    root_nodes: tuple[StructureMapNodeDto, ...]
    edges: tuple[StructureMapEdgeDto, ...]
    filters: StructureMapFiltersDto
    evidence_index: dict[str, tuple[str, ...]] = field(default_factory=dict)
    review_summary: dict[str, Any] = field(default_factory=dict)
    conflict_summary: dict[str, Any] = field(default_factory=dict)
    generated_at: str = ""
    schema: str = STRUCTURE_MAP_PROJECTION_SCHEMA

    def __post_init__(self) -> None:
        if len(self.root_nodes) > self.filters.max_nodes:
            raise ValueError("root_nodes exceeds max_nodes filter")
        if len(self.edges) > self.filters.max_edges:
            raise ValueError("edges exceeds max_edges filter")

File: apps/desktop/src/features/wholeBook/contracts/structureMap.ts
import {
  PATTERN_DTO_HAS_ORM_TABLE,
  STRUCTURE_MAP_DEFAULT_MAX_EDGES,
  STRUCTURE_MAP_DEFAULT_MAX_NODES,
  type StructureMapViewMode,
} from "./keys";

export interface StructureMapNodeDto {
  node_id: string;
  node_type: string;
  title: string;
  asset_id: number | null;
  asset_version_id: number | null;
  is_canonical: boolean;
  view_modes: StructureMapViewMode[];
  chapter_range: [number | null, number | null];
  parent_id: string | null;
  evidence_count: number;
  collapsed: boolean;
  searchable_text: string;
}

export interface StructureMapEdgeDto {
  edge_id: string;
  source_node_id: string;
  target_node_id: string;
  relation_id: number | null;
  relation_version_id: number | null;
  is_canonical: boolean;
  edge_type: string;
  evidence_count: number;
  label: string;
}

export interface StructureMapFiltersDto {
  view_mode: StructureMapViewMode;
  search_query: string;
  include_candidates: boolean;
  max_nodes: number;
  max_edges: number;
  theme: "system" | "light" | "dark";
}

export interface NarrativeStructureMapProjectionDto {
  schema: string;
  book_id: number;
  book_snapshot_id: number;
  source_run_id: number | null;
  projection_version: string;
  root_nodes: StructureMapNodeDto[];
  edges: StructureMapEdgeDto[];
  filters: StructureMapFiltersDto;
  evidence_index: Record<string, string[]>;
  review_summary: Record<string, unknown>;
  conflict_summary: Record<string, unknown>;
  generated_at: string;
}

export const STRUCTURE_MAP_PROJECTION_SCHEMA =
  "narrative_structure_map_projection";

export {
  PATTERN_DTO_HAS_ORM_TABLE,
  STRUCTURE_MAP_DEFAULT_MAX_EDGES,
  STRUCTURE_MAP_DEFAULT_MAX_NODES,
};
